plugins/net_plugin/net_plugin.cpp

```diff
@@ -3136,39 +3134,60 @@ namespace eosio {
+            c->strand.post( [c, id, blk_num=ptr->block_num()]() {
+               my_impl->sync_master->sync_recv_block( c, id, blk_num, false );
+            });
             return;
          }
-         // this may return null if block is not immediately ready to be processed
-         bsp = cc.create_block_state( id, ptr );
-      } catch( const fc::exception& ex ) {
-         exception = true;
-         peer_elog(this, "bad block exception: #${n} ${id}...: ${m}",
-                   ("n", ptr->block_num())("id", id.str().substr(8,16))("m",ex.to_string()));
-      } catch( ... ) {
-         exception = true;
-         peer_elog(this, "bad block: #${n} ${id}...: unknown exception",
-                   ("n", ptr->block_num())("id", id.str().substr(8,16)));
-      }
-      if( exception ) {
-         my_impl->sync_master->rejected_block( shared_from_this(), ptr->block_num() );
-         my_impl->dispatcher->rejected_block( id );
-         return;
-      }
 
-      bool signal_producer = !!bsp; // ready to process immediately, so signal producer to interrupt start_block
-      app().post(priority::medium, [ptr{std::move(ptr)}, bsp{std::move(bsp)}, id, c = shared_from_this()]() mutable {
-         c->process_signed_block( id, std::move(ptr), std::move(bsp) );
-      });
+         block_state_ptr bsp;
+         bool exception = false;
+         try {
+            // this may return null if block is not immediately ready to be processed
+            bsp = cc.create_block_state( id, ptr );
+         } catch( const fc::exception& ex ) {
+            exception = true;
+            fc_elog( logger, "bad block exception connection ${cid}: #${n} ${id}...: ${m}",
+                     ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16))("m",ex.to_string()));
+         } catch( ... ) {
+            exception = true;
+            fc_elog( logger, "bad block connection ${cid}: #${n} ${id}...: unknown exception",
+                     ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16)));
+         }
+         if( exception ) {
+            c->strand.post( [c, id, blk_num=ptr->block_num()]() {
+               my_impl->sync_master->rejected_block( c, blk_num );
+               my_impl->dispatcher->rejected_block( id );
+            });
+            return;
+         }
+
+         bool valid_block_header = !!bsp;
+
+         if( valid_block_header ) {
+            fc_dlog( logger, "validated block header, broadcasting immediately, connection ${cid}, blk num = ${num}, id = ${id}",
+                     ("cid", cid)("num", bsp->block_num)("id", bsp->id) );
+            my_impl->dispatcher->add_peer_block( bsp->id, cid ); // no need to send back to sender
+            my_impl->dispatcher->bcast_block( bsp->block, bsp->id );
+         }
 
-      if( signal_producer )
-         my_impl->producer_plug->received_block();
+         app().post(priority::medium, [ptr{std::move(ptr)}, bsp{std::move(bsp)}, id, c{std::move(c)}]() mutable {
```

Comment:
do we want to do this `post()` if `! bsp`?

Reply:
Yes. `bsp` may be null if the block is not immediately ready to be processed. In that case `producer_plugin::on_incoming_block` will evaluate the block. It will likely be ready by then because of previously posted `process_signed_blocks` have finished ahead of it.

Reply:
Thanks!